source/vir/src/interpreter.rs

```diff
@@ -1521,15 +1572,26 @@ fn eval_expr_launch(
     // Don't run for too long
     let max_iterations = (rlimit as f64 * RLIMIT_MULTIPLIER as f64) as u64 * RLIMIT_MULTIPLIER;
     let ctx = Ctx { fun_ssts: &fun_ssts, max_iterations, arch };
-    let res = eval_expr_internal(&ctx, &mut state, &exp)?;
+    let result = eval_expr_top(&ctx, &mut state, &exp)?;
     display_perf_stats(&state);
     if state.log.is_some() {
         log.replace(state.log.unwrap());
     }
-    if let ExpX::Const(Constant::Bool(false)) = res.x {
-        Err(error(&exp.span, "assert simplifies to false"))
-    } else {
-        match mode {
+
+    match result {
+        SimplificationResult::True => {
+            return Ok((crate::sst_util::sst_bool(&exp.span, true), state.msgs));
+        }
+        SimplificationResult::False(None) => {
+            return Err(error(&exp.span, "expression simplifies to false"));
+        }
+        SimplificationResult::False(Some(small_exp)) => {
```

Comment:
In the `SimplificationResult::Complex(res)` branch, we clean up some of the internal structures we may have created during interpretation.  Will those be leaked in this branch (e.g., what if one of the terms is a sequence expression)?  If so, we might want to hoist the clean-up into its own function and apply it here too.

Reply:
The way the code works pre-PR is that this clean-up happens in the `ComputeMode::Z3` branch, but it doesn't happen in the `ComputeMode::ComputeOnly` where the expression is only used in an error message. After this PR it works the same way. I don't do any cleanup on small_exp, but the only thing it's used for is the error message.

So I don't know if it's intentional, but it shouldn't 'leak' anything unless it was before.

Reply:
Hmm, maybe we should take the opportunity to clean up both error reporting paths, in that case.